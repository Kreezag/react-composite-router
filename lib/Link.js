/* eslint-env browser */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import referringRouter, { statePropTypes } from './referring';
import { router as routerPropTypes } from './propTypes';



class Link extends Component {
  static propTypes = {
    state: statePropTypes,
    target: PropTypes.string,
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
    className: PropTypes.string,
    activeClass: PropTypes.string,
    disabledClass: PropTypes.string,
    activePathClass: PropTypes.string,
    activeStateClass: PropTypes.string
  };

  static contextTypes = {
    router: routerPropTypes.isRequired
  };

  constructor (...args) {
    super(...args);

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick (event) {
    const {
      onClick,
      target,
      disabled,
      state
    } = this.props;

    if (onClick) {
      onClick(event);
    }

    if (
      !event.defaultPrevented
      && !target
      && !(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey) // is modified event
      && event.button === 0 // is left button click
    ) {
      event.preventDefault();

      if (!disabled) {
        state.apply();
      }
    }
  }

  render () {
    const {
      state,
      activeClass,
      activePathClass,
      activeStateClass,
      disabledClass,
      ...otherProps
    } = this.props;

    const className = [
      this.props.className,
      state.isActive && activeClass,
      state.isActivePath && activePathClass,
      state.isActiveState && activeStateClass,
      this.props.disabled && disabledClass
    ].filter(Boolean).join(' ');

    return (
      <a
        {...otherProps}
        href={state.href}
        className={className}
        onClick={this.handleClick}
      />
    );
  }
}


export default referringRouter()(Link);
