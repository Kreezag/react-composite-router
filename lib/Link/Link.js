import { PropTypes } from 'react';
import { compose, getContext } from 'recompose';
import { connect } from 'react-redux';
import LinkComponent from './Link.jsx';
import historyPropTypes from '../util/historyPropTypes';




export default compose(
  getContext({
    history: historyPropTypes(),
    routerRoutesByName: PropTypes.object.isRequired
  }),
  connect(({ routerState }, { history, routerRoutesByName }) => ({
    history,
    routerState,
    routerRoutesByName
  }), {})
)(LinkComponent);
