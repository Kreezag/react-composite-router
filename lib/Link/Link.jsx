/* eslint-env browser */

import React, { PropTypes } from 'react';
import { calcLocation, moveTo } from '../routing';
import historyPropTypes from '../historyPropTypes';


const isLeftClickEvent = (event) =>
event.button === 0;



const isModifiedEvent = (event) =>
  Boolean(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);



const Link = (props) => {
  const {
    routerState,
    routerRoutesByName,
    history,
    state,
    className,
    disabled,
    children,
    reload,
    onClick,
    replace,
    target,
    activeClass,
    activeStateClass,
    disabledClass,
    params = {},
    ...otherProps
  } = props;

  const location = calcLocation(routerRoutesByName, state, routerState.params, params);
  const isActive = history.location.pathname === location.pathname;
  const isActiveState = routerState.name === state || routerState.name.startsWith(state);
  const href = history.createHref(location);

  const handleClick = (event) => {
    if (onClick) {
      onClick(event);
    }

    if (
      !reload
      && !event.defaultPrevented
      && !target
      && !isModifiedEvent(event)
      && isLeftClickEvent(event)
    ) {
      event.preventDefault();

      moveTo(history, location, { replace, reload });
    }
  };

  const classNames = [
    className,
    isActive && activeClass,
    isActiveState && activeStateClass,
    disabled && disabledClass
  ].filter(Boolean).join(' ');

  return (
    <a
      {...otherProps}
      href={href}
      target={target}
      className={classNames}
      disabled={disabled}
      onClick={handleClick}
    >
      {typeof children === 'function' ? children({ isActive, disabled, isActiveState }, routerState) : children}
    </a>
  );
};



Link.propTypes = {
  state: PropTypes.string.isRequired,
  params: PropTypes.object,
  reload: PropTypes.bool.isRequired,
  target: PropTypes.oneOf([ '_blank' ]),
  replace: PropTypes.bool.isRequired,
  onClick: PropTypes.func,
  history: historyPropTypes(),
  disabled: PropTypes.bool.isRequired,
  children: PropTypes.node,
  className: PropTypes.string,
  activeClass: PropTypes.string,
  routerState: PropTypes.object.isRequired,
  disabledClass: PropTypes.string,
  activeStateClass: PropTypes.string,
  routerRoutesByName: PropTypes.object.isRequired
};



Link.defaultProps = {
  reload: false,
  replace: false,
  disabled: false
};



Link.contextTypes = {
};



export default Link;
