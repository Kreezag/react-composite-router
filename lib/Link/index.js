/* eslint-env browser */

import Link from './Link';



export default Link;
