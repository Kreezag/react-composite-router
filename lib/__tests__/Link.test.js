/* eslint-env browser, jest */

jest.mock('../../Router', () => ({
  calcLocation: (state, params) => ({
    pathname: state,
    hash: '',
    search: ''
  })
}));

import React from 'react';
import Link from '../Link.jsx';
import renderer from 'react-test-renderer';
import WithContext from 'react-with-context';
import identity from 'lodash/identity';

describe('rendering', () => {
  it('should render something', () => {
    const component = renderer.create(
      <WithContext context={{
        routerState: { name: 'state', params: {} },
        history: { createHref: identity, location: {} },
        routerRoutesByName: {
          state: { name: 'state' }
        }
      }}>
        <Link to="state">Hello!</Link>
      </WithContext>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
