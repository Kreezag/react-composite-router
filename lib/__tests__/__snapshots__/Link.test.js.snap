exports[`Link should render something 1`] = `
<a
  className="LinkClass active active-state disabled"
  disabled={true}
  href="/"
  onClick={[Function]}
  target="_blank">
  Hello!
</a>
`;
